In [ ]:
import pandas as pd
from datetime import datetime, timedelta

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# Load the 3-month dataset
file_path_3m = '/content/drive/MyDrive/DSB/UseCase_1/Data/3m_data.csv'
data_3m = pd.read_csv(file_path_3m)

# Display basic information about the dataset and check the first few rows
data_3m_info = data_3m.info()
data_3m_head = data_3m.head()
data_3m_info, data_3m_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7014 entries, 0 to 7013
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           7014 non-null   int64  
 1   ID                   7014 non-null   float64
 2   Reached_3w           7014 non-null   float64
 3   Reached_3m           7014 non-null   float64
 4   Tenure_In_Months     7014 non-null   float64
 5   Date_Joined          7014 non-null   object 
 6   FUA_Balance          7014 non-null   float64
 7   Number_Of_Services   7014 non-null   float64
 8   Has_Payroll          7014 non-null   float64
 9   Income               508 non-null    float64
 10  Has_Investment       7014 non-null   float64
 11  Has_Visa             7014 non-null   float64
 12  VISA_balance         7014 non-null   float64
 13  Has_mortgage         7014 non-null   float64
 14  Has_Deposit          7014 non-null   float64
 15  Number_Transactions  7014 non-null   f

(None,
    Unnamed: 0      ID  Reached_3w  Reached_3m  Tenure_In_Months Date_Joined  \
 0           0  2613.0         0.0         0.0               3.0   7/15/2020   
 1           1  3382.0         1.0         1.0               3.0   1/21/2021   
 2           2  3927.0         1.0         1.0               3.0   8/14/2020   
 3           3  4065.0         1.0         1.0               3.0   6/26/2020   
 4           4  4221.0         0.0         0.0               3.0  12/18/2020   
 
    FUA_Balance  Number_Of_Services  Has_Payroll  Income  Has_Investment  \
 0       944.35                 2.0          1.0     NaN             0.0   
 1      1589.61                 2.0          0.0     NaN             0.0   
 2      1000.00                 3.0          1.0     NaN             0.0   
 3       219.41                 1.0          0.0     NaN             0.0   
 4        76.09                 1.0          0.0     NaN             0.0   
 
    Has_Visa  VISA_balance  Has_mortgage  Has_Deposit

In [ ]:
# Check for missing values in the dataset
missing_values_3m = data_3m.isnull().sum()
missing_values_3m

Unnamed: 0                0
ID                        0
Reached_3w                0
Reached_3m                0
Tenure_In_Months          0
Date_Joined               0
FUA_Balance               0
Number_Of_Services        0
Has_Payroll               0
Income                 6506
Has_Investment            0
Has_Visa                  0
VISA_balance              0
Has_mortgage              0
Has_Deposit               0
Number_Transactions       0
Age                       0
City                      0
Unnamed: 17            7014
dtype: int64

From the initial inspection of the 3-month dataset:

* The dataset has 7014 entries.
* There is a substantial amount of missing data in the Income field, with 6506 missing values.
* An empty column, Unnamed: 17, exists and should be removed.
* The City field seems to have extra spaces; cleaning this field might be necessary.
* The Date_Joined field is of object (string) type; converting it to datetime will be beneficial for any time-related analysis.

In [ ]:
# Drop 'Unnamed: 17' (empty column) & Unnamed: 0' (not needed)
data_3m.drop(columns=['Unnamed: 17', 'Unnamed: 0'], inplace=True)

In [ ]:
# Strip extra spaces from 'City' field
data_3m['City'] = data_3m['City'].str.strip()

In [ ]:
# Convert 'Date_Joined' to datetime
data_3m['Date_Joined'] = pd.to_datetime(data_3m['Date_Joined'])

In [ ]:
# Check the cleaned dataset
data_3m.head()

ID  Reached_3w  Reached_3m  Tenure_In_Months Date_Joined  FUA_Balance  \
0  2613.0         0.0         0.0               3.0  2020-07-15       944.35   
1  3382.0         1.0         1.0               3.0  2021-01-21      1589.61   
2  3927.0         1.0         1.0               3.0  2020-08-14      1000.00   
3  4065.0         1.0         1.0               3.0  2020-06-26       219.41   
4  4221.0         0.0         0.0               3.0  2020-12-18        76.09   

   Number_Of_Services  Has_Payroll  Income  Has_Investment  Has_Visa  \
0                 2.0          1.0     NaN             0.0       0.0   
1                 2.0          0.0     NaN             0.0       1.0   
2                 3.0          1.0     NaN             0.0       1.0   
3                 1.0          0.0     NaN             0.0       1.0   
4                 1.0          0.0     NaN             0.0       1.0   

   VISA_balance  Has_mortgage  Has_Deposit  Number_Transactions   Age  \
0      0.000000           0.0          1.0                 45.0  49.0   
1    676.347873           0.0          1.0                 25.0  69.0   
2    326.556766           0.0          1.0                  0.0  59.0   
3      0.000000           0.0          1.0                  0.0  71.0   
4   1315.930929           0.0          1.0                 66.0  65.0   

            City  
0         SURREY  
1  PENDER ISLAND  
2        BURNABY  
3      VANCOUVER  
4      VANCOUVER

In [ ]:
data_3m.to_csv('/content/drive/MyDrive/DSB/UseCase_1/Data/3m_cdata.csv', index=False)

In [ ]:
# Load the 6-month dataset
file_path_6m = '/content/drive/MyDrive/DSB/UseCase_1/Data/6m_data.csv'
data_6m = pd.read_csv(file_path_6m)

# Display basic information about the dataset and check the first few rows
data_6m_info = data_6m.info()
data_6m_head = data_6m.head()
data_6m_info, data_6m_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6023 entries, 0 to 6022
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   6023 non-null   int64  
 1   Reached_3w           6023 non-null   int64  
 2   Reached_3m           6023 non-null   int64  
 3   Tenure_In_Months     6023 non-null   int64  
 4   Date_Joined          6023 non-null   object 
 5   FUA_Balance          6023 non-null   float64
 6   Number_Of_Services   6023 non-null   int64  
 7   Has_Payroll          6023 non-null   int64  
 8   Income               431 non-null    float64
 9   Has_Investment       6023 non-null   int64  
 10  Has_Visa             6023 non-null   int64  
 11  VISA_balance         6023 non-null   float64
 12  Has_mortgage         6023 non-null   int64  
 13  Has_Deposit          6023 non-null   int64  
 14  Number_Transactions  6023 non-null   int64  
 15  Age                  6023 non-null   i

(None,
         ID  Reached_3w  Reached_3m  Tenure_In_Months Date_Joined  \
 0  9556672           0           1                 6   6/20/2020   
 1  7671094           0           0                 6   7/29/2020   
 2  5261655           0           0                 6   8/10/2020   
 3  8042348           0           0                 6   6/18/2020   
 4  5159745           1           0                 6    2/6/2021   
 
     FUA_Balance  Number_Of_Services  Has_Payroll  Income  Has_Investment  \
 0   49907.23505                   6            0     NaN               0   
 1   10702.44980                   2            0     NaN               0   
 2  704362.97720                   1            0     NaN               0   
 3  413309.42910                   3            1     NaN               1   
 4   41268.42368                   1            1     NaN               1   
 
    Has_Visa  VISA_balance  Has_mortgage  Has_Deposit  Number_Transactions  \
 0         0           0.0         

In [ ]:
# Check for missing values in the dataset
missing_values_6m = data_6m.isnull().sum()
missing_values_6m

ID                        0
Reached_3w                0
Reached_3m                0
Tenure_In_Months          0
Date_Joined               0
FUA_Balance               0
Number_Of_Services        0
Has_Payroll               0
Income                 5592
Has_Investment            0
Has_Visa                  0
VISA_balance              0
Has_mortgage              0
Has_Deposit               0
Number_Transactions       0
Age                       0
City                      0
dtype: int64

From the initial inspection of the 6-month dataset:

* The dataset has 6023 entries.
* Similar to the 3-month dataset, the Income field has a substantial amount of missing data, with 5592 missing values.
* The City field seems to have extra spaces; cleaning this field might be necessary.
* The Date_Joined field is of object (string) type; converting it to datetime will be beneficial for any time-related analysis.

In [ ]:
# Strip extra spaces from 'City' field
data_6m['City'] = data_6m['City'].str.strip()

In [ ]:
# Convert 'Date_Joined' to datetime
# data_6m['Date_Joined'] = pd.to_datetime(data_6m['Date_Joined'])

ParserError: ignored

In [ ]:
# Extract unique values in the 'Date_Joined' column of the 6-month dataset to inspect the discrepancies
unique_dates_6m = data_6m['Date_Joined'].unique()
unique_dates_6m

array(['6/20/2020', '7/29/2020', '8/10/2020', '6/18/2020', '2/6/2021',
       '2/4/2021', '9/11/2020', '8/5/2020', '1/13/2021', '12/3/2020',
       '5/8/2020', '4/1/2020', '6/9/2020', '4/26/2020', '4/16/2020',
       '1/5/2021', '4/25/2020', '6/8/2020', '7/14/2020', '9/9/2020',
       '7/21/2020', '2/10/2021', '4/20/2020', '7/31/2020', '8/28/2020',
       '4/8/2020', '7/13/2020', '9/2/2020', '12/4/2020', '1/21/2021',
       '5/28/2020', '8/13/2020', '2/3/2021', '5/5/2020', '6/11/2020',
       '4/7/2020', '7/22/2020', '5/12/2020', '3/20/2020', '8/27/2020',
       '8/22/2020', '2/13/2021', '1/6/2021', '1/7/2021', '8/29/2020',
       '12/24/2020', '5/7/2020', '9/10/2020', '12/11/2020', '8/14/2020',
       '4/21/2020', '7/10/2020', '3/23/2020', '6/24/2020', '4/3/2020',
       '12/20/2020', '7/24/2020', '5/1/2020', '4/23/2020', '8/19/2020',
       '8/7/2020', '12/12/2020', '1/15/2021', '12/16/2020', '7/28/2020',
       '1/14/2021', '3/30/2020', '3/21/2020', '12/5/2020', '1/16/2021',
       

In [ ]:
def convert_excel_date_to_datetime(excel_date):
    return datetime(1899, 12, 30) + timedelta(days=int(excel_date))

def convert_date_joined_to_datetime(date_joined):
    try:
        return datetime.strptime(date_joined, '%m/%d/%Y')
    except ValueError:
        try:
            return convert_excel_date_to_datetime(date_joined)
        except:
            return date_joined

data_6m['Date_Joined'] = data_6m['Date_Joined'].apply(convert_date_joined_to_datetime)

In [ ]:
unique_dates_6m_converted = data_6m['Date_Joined'].unique()
unique_dates_6m_converted

array(['2020-06-20T00:00:00.000000000', '2020-07-29T00:00:00.000000000',
       '2020-08-10T00:00:00.000000000', '2020-06-18T00:00:00.000000000',
       '2021-02-06T00:00:00.000000000', '2021-02-04T00:00:00.000000000',
       '2020-09-11T00:00:00.000000000', '2020-08-05T00:00:00.000000000',
       '2021-01-13T00:00:00.000000000', '2020-12-03T00:00:00.000000000',
       '2020-05-08T00:00:00.000000000', '2020-04-01T00:00:00.000000000',
       '2020-06-09T00:00:00.000000000', '2020-04-26T00:00:00.000000000',
       '2020-04-16T00:00:00.000000000', '2021-01-05T00:00:00.000000000',
       '2020-04-25T00:00:00.000000000', '2020-06-08T00:00:00.000000000',
       '2020-07-14T00:00:00.000000000', '2020-09-09T00:00:00.000000000',
       '2020-07-21T00:00:00.000000000', '2021-02-10T00:00:00.000000000',
       '2020-04-20T00:00:00.000000000', '2020-07-31T00:00:00.000000000',
       '2020-08-28T00:00:00.000000000', '2020-04-08T00:00:00.000000000',
       '2020-07-13T00:00:00.000000000', '2020-09-02

In [ ]:
# Check the cleaned dataset
data_6m.head()

ID  Reached_3w  Reached_3m  Tenure_In_Months Date_Joined  \
0  9556672           0           1                 6  2020-06-20   
1  7671094           0           0                 6  2020-07-29   
2  5261655           0           0                 6  2020-08-10   
3  8042348           0           0                 6  2020-06-18   
4  5159745           1           0                 6  2021-02-06   

    FUA_Balance  Number_Of_Services  Has_Payroll  Income  Has_Investment  \
0   49907.23505                   6            0     NaN               0   
1   10702.44980                   2            0     NaN               0   
2  704362.97720                   1            0     NaN               0   
3  413309.42910                   3            1     NaN               1   
4   41268.42368                   1            1     NaN               1   

   Has_Visa  VISA_balance  Has_mortgage  Has_Deposit  Number_Transactions  \
0         0           0.0             0            1                    2   
1         1           0.0             0            1                   20   
2         0           0.0             0            1                    0   
3         0           0.0             0            1                    0   
4         0           0.0             0            1                    0   

   Age             City  
0   54          LANGLEY  
1   68        VANCOUVER  
2   48  NORTH VANCOUVER  
3   67        COQUITLAM  
4   60        VANCOUVER

In [ ]:
data_6m.to_csv('/content/drive/MyDrive/DSB/UseCase_1/Data/6m_cdata.csv', index=False)